### PR TITLE
feat(core): 🎸 overlay

### PR DESCRIPTION
**Pull request by wjian23**
## Purpose of this Pull Request

new feature overlay support.

## Checklist Before Merging

Please ensure the following are completed before merging:
- [x] If this is linked to an issue, include the link in your description.
- [x] If you've made changes to the code or documentation, make sure these are updated in the `CHANGELOG.md` file.
- [x] If you've introduced any break changes, briefly describe them in the `Breaking` section of the `CHANGELOG.md` file.

## Additional Information

If you're unsure about which branch to submit your Pull Request to, or when it will be released after being merged, please refer to our [Release Guide](https://github.com/RibirX/Ribir/blob/master/RELEASE.md).

If you're working on a widget and need help writing test cases, we have some macros that can assist you. Please refer to the [Ribir Dev Helper](https://docs.rs/ribir_dev_helper) documentation.

### CODE REVIEW

File: core/src/window.rs [hunk 1]
```diff
@@ -146,6 +146,39 @@ impl Window {
       .focus_next_widget(&self.widget_tree.borrow().arena);
   }
 
+  /// Insert a overlay widget to the window and return the control of the
```
Comment (M-Adoo) [inline core/src/window.rs:149]: I'd like to use an overlay like this:

```rust


fn_widget! {
  let menu = overlay! {
    @List {
      ...
    }
  };

  @Button {
    on_tap: move |_| overlay.show(ctx!()),
    ...
  }
}
```

We can use the `overlay` macro to create an overlay widget. This macro only requires the `ctx` (context) when the overlay is displayed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -28,6 +28,9 @@ Please only add new entries below the [Unreleased](#unreleased---releasedate) he
 
 - fix broken links and format the example code (#526 @M-Adoo)
 
+### Features
+- Support the overlay (@wjian23)
```
Comment (M-Adoo) [inline CHANGELOG.md:32]: As a so important feature, we should provide a more detailed introduction to the overlay.

And I think we should write a chapter about it and put it in `docs/en/advanced_topics/overlay.md`.

File: core/src/overlay.rs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+  fn map<O>(mut self, f: impl Fn(Writer<OverlayControl>, Self::Value) -> O) -> impl OverlayBuilder
+  where
+    O: WidgetBuilder + 'static,
+  {
+    OverlayCreator {
+      builder: move |ctrl: Writer<OverlayControl>| f(ctrl.clone_writer(), self.build(ctrl)),
+    }
+  }
+
+  fn with_mask_layer(self, brush: impl Into<Brush>, tap_remove: bool) -> impl OverlayBuilder
```
Comment (M-Adoo) [inline core/src/overlay.rs:50]: We should provide a style to make the overlay more flexible, not just a mask layer. I prefer to remove the `with_mask_layer` and support the style when we refactor the widgets library.

File: core/src/widget_tree.rs [hunk 4]
```diff
@@ -28,8 +28,8 @@ pub(crate) struct WidgetTree {
 impl WidgetTree {
   pub fn init(&mut self, wnd: Weak<Window>) { self.wnd = wnd; }
 
-  pub fn set_root(&mut self, root: WidgetId) {
-    self.root = root;
+  pub fn set_content(&mut self, content: WidgetId) {
+    self.root.append(content, &mut self.arena);
```
Comment (M-Adoo) [inline core/src/widget_tree.rs:32]: What happen if we call `set_content` twice with different widget?

File: core/src/overlay.rs [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+
+/// Use to control the overlay
+pub trait OverlayControl: OverlayBuilder {
+  /// the Overlay widget will be show at the top level of all widget.
+  /// if the overlay is showing, nothing will happen.
+  fn show(&self, wnd: Rc<Window>);
+
+  /// User can make transform before the widget show at the top level of all
+  /// widget. if the overlay is showing, nothing will happen.
+  /// ### Example
```
Comment (M-Adoo) [inline core/src/overlay.rs:53]: Insert a blank line before the title for readability

File: core/src/overlay.rs [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+where
+  M: Fn(OverlayCloseHandle) -> O + 'static,
+  O: WidgetBuilder + 'static,
+{
+  type Output = O;
+  fn build(&self, ctrl: OverlayCloseHandle) -> Self::Output { (self.0.builder)(ctrl) }
+}
+
+#[derive(Clone)]
+pub struct OverlayWidget<M>(M);
```
Comment (M-Adoo) [inline core/src/overlay.rs:307]: The inner type does not need to be public

File: core/src/overlay.rs [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+      },
+      wnd,
+    );
+  }
+
+  /// return whether the overlay is show.
+  fn is_show(&self) -> bool;
+
+  /// remove the showing overlay.
+  fn remove(&self);
```
Comment (M-Adoo) [inline core/src/overlay.rs:117]: Consider using `close` or `exit` instead of `remove` for better clarity.

File: core/src/overlay.rs [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+
+/// A handle to close the overlay
+#[derive(Clone)]
+pub struct OverlayCloseHandle(OverlayState);
+impl OverlayCloseHandle {
+  pub fn remove(&self) { self.0.remove() }
+}
+
+/// Use to control the overlay
+pub trait OverlayControl: OverlayBuilder {
```
Comment (M-Adoo) [inline core/src/overlay.rs:46]: I think `OverlayControl` is unnecessary, directly implement methods for `Overlay` is better.

File: core/src/widget_tree.rs (hunk 4 above)
Comment (wjian23) [reply core/src/widget_tree.rs:39]: yeah, here should clean the old content

File: core/src/overlay.rs [hunk 9]
```diff
@@ -0,0 +1,426 @@
...
+  ///   );
+  ///   @FilledButton{
+  ///     on_tap: move |e| overlay.show(e.window()),
+  ///     @{ Label::new("Click to show overlay") }
+  ///   }
+  ///  };
+  ///  App::new_window(w, None);
+  ///  App::exec();
+  /// ```
+  pub fn from_widget(w: M, ctx: &BuildCtx) -> Self {
```
Comment (M-Adoo) [inline core/src/overlay.rs:207]: Consider not requiring a `ctx` parameter here. Instead, query the default style in the `show` method.

File: core/src/overlay.rs (hunk 6 above)
Comment (wjian23) [reply core/src/overlay.rs:310]: the type is refed by a pub method from_widget as return type

File: CHANGELOG.md [hunk 10]
```diff
@@ -22,6 +22,10 @@ Please only add new entries below the [Unreleased](#unreleased---releasedate) he
 
 ## [@Unreleased] - @ReleaseDate
 
+### Features
+
+- Support the overlay (@wjian23)
```
Comment (M-Adoo) [inline CHANGELOG.md:27]: Introduced the Overlay Feature: This enhancement simplifies the creation of overlay widgets. It streamlines the addition of any widget to an overlay and offers a more user-friendly API for overlay management

File: CHANGELOG.md (hunk 2 above)
Comment (wjian23) [reply CHANGELOG.md]: have put more example to rust doc.

**Review by M-Adoo — APPROVED** [review]

**Review by M-Adoo — APPROVED** [review]

**Review by sologeek — APPROVED** [review]

**Pull request merged**